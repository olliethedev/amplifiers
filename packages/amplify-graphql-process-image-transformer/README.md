# amplify-graphql-process-image-transformer #

## Description ##

WIP

## Development and Contributions ##
Contributions are more than welcome! Please feel free to open an issue or a pull request.
Developer docs are [here](https://github.com/olliethedev/amplifiers)

## License ##
[MIT License](https://github.com/olliethedev/amplifiers/blob/master/LICENSE)