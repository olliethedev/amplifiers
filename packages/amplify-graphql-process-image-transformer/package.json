{
  "name": "@amplifiers/amplify-graphql-process-image-transformer",
  "version": "1.1.13",
  "description": "This directive allows you to process images.",
  "keywords": [
    "aws",
    "amplify",
    "grapqhl",
    "serverless",
    "transformer",
    "dynamodb",
    "image",
    "s3",
    "processing"
  ],
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "dependencies": {
    "@amplifiers/amplify-graphql-amplifiers-core": "^1.0.0",
    "@aws-amplify/graphql-transformer-core": "0.17.11",
    "@aws-amplify/graphql-transformer-interfaces": "1.14.7",
    "@aws-cdk/aws-appsync": "~1.124.0",
    "@aws-cdk/aws-iam": "~1.124.0",
    "@aws-cdk/aws-lambda": "~1.124.0",
    "@aws-cdk/core": "~1.124.0",
    "graphql-mapping-template": "4.20.5",
    "graphql-transformer-common": "4.24.0"
  },
  "devDependencies": {
    "@aws-amplify/graphql-model-transformer": "0.15.6",
    "@aws-cdk/assert": "~1.124.0",
    "@types/node": "^20.8.9",
    "aws-sdk": "^2.1090.0",
    "bestzip": "^2.1.5",
    "graphql": "^14.5.8",
    "mkdirp": "^0.5.5"
  },
  "scripts": {
    "compile": "rm -rf dist/* && tsc -p tsconfig.pkg.json && mkdirp dist/assets && mkdirp dist/assets && cd resources/lambda-functions && npm install && npm run fix_sharp_aws && bestzip --force node ../../dist/assets/lambda.zip index.js node_modules/",
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "author": "olliethedev",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/olliethedev/amplifiers/issues"
  },
  "homepage": "https://github.com/olliethedev/amplifiers#readme",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/olliethedev/amplifiers.git",
    "directory": "packages/amplify-graphql-process-image-transformer"
  }
}