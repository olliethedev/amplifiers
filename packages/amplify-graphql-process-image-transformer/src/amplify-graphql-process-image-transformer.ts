import {
    DirectiveWrapper,
    InvalidDirectiveError,
    TransformerPluginBase,
} from '@aws-amplify/graphql-transformer-core';
import { ModelResourceIDs, ResourceConstants } from "graphql-transformer-common";
import {
    TransformerContextProvider,
} from '@aws-amplify/graphql-transformer-interfaces';
import { TransformerSchemaVisitStepContextProvider } from '@aws-amplify/graphql-transformer-interfaces/src';
import { DirectiveNode, FieldDefinitionNode, InterfaceTypeDefinitionNode, ObjectTypeDefinitionNode } from 'graphql';
import { createEventSourceMapping, createLambda } from './create-process-image-lambda';
import { Table } from '@aws-cdk/aws-dynamodb';
import { DynamoDbDataSource } from '@aws-cdk/aws-appsync';
import { IConstruct, CfnCondition, Fn, CfnParameter } from '@aws-cdk/core';
import { IFunction } from '@aws-cdk/aws-lambda';
import { IRole, Role, ServicePrincipal } from '@aws-cdk/aws-iam';
import { DirectiveArgs, ProcessorAction } from './directive-type';

const STACK_NAME = 'ProcessImageTransformer';

const directiveName = "processImage";



interface DirectiveObjectTypeDefinition {
    node: ObjectTypeDefinitionNode;
    fieldName: string;
    fieldParams: DirectiveArgs;
}

export class Transformer extends TransformerPluginBase {
    private directiveMap: Record<string, DirectiveObjectTypeDefinition> = {};

    constructor() {
        super(
            'amplify-graphql-process-image-transformer',
        /* GraphQL */ `
          directive @${ directiveName }(bucket:String! actions:[AWSJSON!]! ) on FIELD_DEFINITION
        `,
        );
    }

    field = (
        parent: ObjectTypeDefinitionNode | InterfaceTypeDefinitionNode,
        definition: FieldDefinitionNode,
        directive: DirectiveNode,
        context: TransformerSchemaVisitStepContextProvider,
    ): void => {
        validateModelDirective(parent as ObjectTypeDefinitionNode);
        validateFieldDirective(definition);
        const directiveArguments = getDirectiveArguments(directive);
        validateDirectiveArguments(directiveArguments, definition);

        this.directiveMap[definition.name.value] = {
            node: parent as ObjectTypeDefinitionNode,
            fieldName: definition.name.value,
            fieldParams: directiveArguments,
        };
    }

    generateResolvers = (context: TransformerContextProvider): void => {
        const stack = context.stackManager.createStack(STACK_NAME);
        const env = context.stackManager.getParameter(ResourceConstants.PARAMETERS.Env) as CfnParameter;

        stack.templateOptions.templateFormatVersion = '2010-09-09';
        stack.templateOptions.description = `An auto-generated nested stack for the @${ directiveName } directive.`;

        new CfnCondition(stack, ResourceConstants.CONDITIONS.HasEnvironmentParameter, {
            expression: Fn.conditionNot(Fn.conditionEquals(env, ResourceConstants.NONE)),
        });

        // lambda role
        const role = new Role(stack, `${ STACK_NAME }LambdaRole`, {
            assumedBy: new ServicePrincipal('lambda.amazonaws.com'),
        });

        // create lambda inputs
        const fieldMappings = Object.keys(this.directiveMap).reduce((acc, modelName) => {
            const { fieldName, fieldParams, node } = this.directiveMap[modelName];
            const table = getTable(context, node);
            const ddbTable = table as Table;
            acc[ddbTable.tableArn] = { 
                fieldName, 
                args: fieldParams
            };
            return acc;
        }, {} as Record<string, { fieldName: string, args: DirectiveArgs }>);

        // create lambda function
        const lambda = createLambda(
            stack, context, role, fieldMappings
        );

        // creates event source mapping for each table
        createSourceMappings(this.directiveMap, context, lambda);
    };
}

const validateModelDirective = (object: ObjectTypeDefinitionNode): void => {
    const modelDirective = object.directives!.find((dir) => dir.name.value === 'model');
    if (!modelDirective) {
        throw new InvalidDirectiveError('The @default directive may only be added to object definitions annotated with @model.');
    }
};

const validateFieldDirective = (field: FieldDefinitionNode): void => {
    if (field.type.kind === 'NonNullType' && (field as any).type.type.name.value !== 'AWSURL') {
        throw new InvalidDirectiveError(`@${ directiveName } can only be used on a field of type AWSURL.`);
    }
    if (field.type.kind === 'NamedType' && field.type.name.value !== 'AWSURL') {
        throw new InvalidDirectiveError(`@${ directiveName } can only be used on a field of type AWSURL.`);
    }
};

const getDirectiveArguments = (directive: DirectiveNode): DirectiveArgs => {
    const directiveWrapped = new DirectiveWrapper(directive);
    return directiveWrapped.getArguments<DirectiveArgs>({
        bucket: "",
        actions: [
            {
                type: "resize",
                name: "small",
                width: 100,
                height: 100,
            } as ProcessorAction
        ]
    });
}

const validateDirectiveArguments = (directiveArguments: DirectiveArgs, field: FieldDefinitionNode): void => {
    if (directiveArguments.actions.length === 0) throw new InvalidDirectiveError(`@${ directiveName } requires an actions argument to be an array with at least one item.`);
    directiveArguments.actions.forEach((action, index) => {
        if (!action.type || action.type.length === 0) throw new InvalidDirectiveError(`@${ directiveName } requires an actions argument to have a type property.`);
        if (!["resize"].includes(action.type)) throw new InvalidDirectiveError(`@${ directiveName } requires an actions argument to have a type property with a value of "resize".`);
        if (!action.name || action.name.length === 0) throw new InvalidDirectiveError(`@${ directiveName } requires an actions argument to have a name property.`);
        if (!action.width || isNaN(action.width)) throw new InvalidDirectiveError(`@${ directiveName } requires an actions argument to have a width property that is an integer.`);
        if (!action.height || isNaN(action.height)) throw new InvalidDirectiveError(`@${ directiveName } requires an actions argument to have a height property.`);
    });
}

const createSourceMappings = (directiveMap: Record<string, DirectiveObjectTypeDefinition>, context: TransformerContextProvider, lambda: IFunction): void => {
    const stack = context.stackManager.getStack(STACK_NAME);
    for (const modelName of Object.keys(directiveMap)) {
        const def = directiveMap[modelName];
        const type = def.node.name.value;
        const indexName = context.resourceHelper.getModelNameMapping(type);
        const table = getTable(context, def.node);
        const ddbTable = table as Table;

        ddbTable.grantStreamRead(lambda.role as IRole);

        createEventSourceMapping(stack, indexName, lambda, ddbTable.tableStreamArn as string);
    }
}

const getTable = (context: TransformerContextProvider, definition: ObjectTypeDefinitionNode): IConstruct => {
    const ddbDataSource = context.dataSources.get(definition) as DynamoDbDataSource;
    const tableName = ModelResourceIDs.ModelTableResourceID(definition.name.value);
    const table = ddbDataSource.ds.stack.node.findChild(tableName);
    return table;
};