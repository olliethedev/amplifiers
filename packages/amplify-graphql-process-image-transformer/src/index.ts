import { Transformer } from "./amplify-graphql-process-image-transformer";

export default Transformer;