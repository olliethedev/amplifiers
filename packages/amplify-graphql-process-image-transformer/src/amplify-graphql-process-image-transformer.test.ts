import {  GraphQLTransform } from '@aws-amplify/graphql-transformer-core';
import { ModelTransformer } from '@aws-amplify/graphql-model-transformer';
import { parse } from 'graphql';
import { Transformer } from './amplify-graphql-process-image-transformer';


test('Process Image Transformer validation happy case', () => {
    const validSchema = `
    type Todo
    @model {
    id: ID!
    picture: AWSURL @processImage(
        bucket: "appimagestoreage113252-dev"
        actions: [{type:"resize", name:"thumbnail",  width:100, height:100},
        {name:"medium", type:"resize", width:500, height:500}]
      )
  }`;
    const transformer = new GraphQLTransform({
      transformers: [new ModelTransformer(), new Transformer()],
    });
    const out = transformer.transform(validSchema);
    expect(out).toBeDefined();
    parse(out.schema);
    expect(out.schema).toMatchSnapshot();
  });

  test('Not creating Process Image Transformer without @model', () => {
    const invalidSchema = `
    type Todo {
        id: ID!
        picture: AWSURL @processImage(
            bucket: "appimagestoreage113252-dev"
            actions: [{type:"resize", name:"thumbnail",  width:100, height:100},
            {name:"medium", type:"resize", width:500, height:500}]
        ) 
  }`;
    const transformer = new GraphQLTransform({
      transformers: [new ModelTransformer(), new Transformer()],
    });
    expect(() => transformer.transform(invalidSchema)).toThrowError();
  })

  test('Not creating Process Image Transformer without AWSURL!', () => {
    const invalidSchema = `
    type Todo 
        @model{
        id: ID!
        picture: String! @processImage(
            bucket: "appimagestoreage113252-dev"
            actions: [{type:"resize", name:"thumbnail",  width:100, height:100},
            {name:"medium", type:"resize", width:500, height:500}]
        ) 
  }`;
    const transformer = new GraphQLTransform({
      transformers: [new ModelTransformer(), new Transformer()],
    });
    expect(() => transformer.transform(invalidSchema)).toThrowError();
  })

  test('Not creating Process Image Transformer without AWSURL', () => {
    const invalidSchema = `
    type Todo 
        @model{
        id: ID!
        picture: String @processImage(
            bucket: "appimagestoreage113252-dev"
            actions: [{type:"resize", name:"thumbnail",  width:100, height:100},
            {name:"medium", type:"resize", width:500, height:500}]
        ) 
  }`;
    const transformer = new GraphQLTransform({
      transformers: [new ModelTransformer(), new Transformer()],
    });
    expect(() => transformer.transform(invalidSchema)).toThrowError();
  })

  test('Not creating Process Image Transformer without bucket argument', () => {
    const invalidSchema = `
    type Todo 
        @model{
        id: ID!
        picture: AWSURL @processImage
  }`;
    const transformer = new GraphQLTransform({
      transformers: [new ModelTransformer(), new Transformer()],
    });
    expect(() => transformer.transform(invalidSchema)).toThrowError();
  })

  test('Not creating Process Image Transformer without valid type argument', () => {
    const invalidSchema = `
    type Todo 
        @model{
        id: ID!
        picture: AWSURL @processImage(
            bucket: "appimagestoreage113252-dev"
            actions: [{ name:"thumbnail",  width:100, height:100},
            {name:"medium", type:"resize", width:500, height:500}]
        ) 
  }`;
    const transformer = new GraphQLTransform({
      transformers: [new ModelTransformer(), new Transformer()],
    });
    expect(() => transformer.transform(invalidSchema)).toThrowError();
  })

  test('Not creating Process Image Transformer without valid actions argument', () => {
    const invalidSchema = `
    type Todo 
        @model{
        id: ID!
        picture: AWSURL @processImage(
            bucket: "appimagestoreage113252-dev"
            actions: []
        ) 
  }`;
    const transformer = new GraphQLTransform({
      transformers: [new ModelTransformer(), new Transformer()],
    });
    expect(() => transformer.transform(invalidSchema)).toThrowError();
  })

  test('Not creating Process Image Transformer without valid action type', () => {
    const invalidSchema = `
    type Todo 
        @model{
        id: ID!
        picture: AWSURL @processImage(
            bucket: "appimagestoreage113252-dev"
            actions: [{ name:"thumbnail", type:"foobar", width:100, height:100},
            {name:"medium", type:"foobar", width:500, height:500}]
        ) 
  }`;
    const transformer = new GraphQLTransform({
      transformers: [new ModelTransformer(), new Transformer()],
    });
    expect(() => transformer.transform(invalidSchema)).toThrowError();
  })

  test('Not creating Process Image Transformer without valid action name', () => {
    const invalidSchema = `
    type Todo 
        @model{
        id: ID!
        picture: AWSURL @processImage(
            bucket: "appimagestoreage113252-dev"
            actions: [{ name:"thumbnail", type:"resize", width:100, height:100},
            {name:"", type:"resize", width:500, height:500}]
        ) 
  }`;
    const transformer = new GraphQLTransform({
      transformers: [new ModelTransformer(), new Transformer()],
    });
    expect(() => transformer.transform(invalidSchema)).toThrowError();
  })

  test('Not creating Process Image Transformer without valid action width', () => {
    const invalidSchema = `
    type Todo 
        @model{
        id: ID!
        picture: AWSURL @processImage(
            bucket: "appimagestoreage113252-dev"
            actions: [{ name:"thumbnail", type:"resize", width:100, height:100},
            {name:"thumb", type:"resize", width:"500px", height:500}]
        ) 
  }`;
    const transformer = new GraphQLTransform({
      transformers: [new ModelTransformer(), new Transformer()],
    });
    expect(() => transformer.transform(invalidSchema)).toThrowError();
  })

  test('Not creating Process Image Transformer without valid action height', () => {
    const invalidSchema = `
    type Todo 
        @model{
        id: ID!
        picture: AWSURL @processImage(
            bucket: "appimagestoreage113252-dev"
            actions: [{ name:"thumbnail", type:"resize", width:100, height:100},
            {name:"thumb", type:"resize", width:500, height:"500px"}]
        ) 
  }`;
    const transformer = new GraphQLTransform({
      transformers: [new ModelTransformer(), new Transformer()],
    });
    expect(() => transformer.transform(invalidSchema)).toThrowError();
  })