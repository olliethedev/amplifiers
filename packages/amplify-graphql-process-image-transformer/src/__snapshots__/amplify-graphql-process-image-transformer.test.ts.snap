// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Process Image Transformer validation happy case 1`] = `
"
type Todo {
  id: ID!
  picture: AWSURL
  createdAt: AWSDateTime!
  updatedAt: AWSDateTime!
}

input ModelStringInput {
  ne: String
  eq: String
  le: String
  lt: String
  ge: String
  gt: String
  contains: String
  notContains: String
  between: [String]
  beginsWith: String
  attributeExists: Boolean
  attributeType: ModelAttributeTypes
  size: ModelSizeInput
}

input ModelIntInput {
  ne: Int
  eq: Int
  le: Int
  lt: Int
  ge: Int
  gt: Int
  between: [Int]
  attributeExists: Boolean
  attributeType: ModelAttributeTypes
}

input ModelFloatInput {
  ne: Float
  eq: Float
  le: Float
  lt: Float
  ge: Float
  gt: Float
  between: [Float]
  attributeExists: Boolean
  attributeType: ModelAttributeTypes
}

input ModelBooleanInput {
  ne: Boolean
  eq: Boolean
  attributeExists: Boolean
  attributeType: ModelAttributeTypes
}

input ModelIDInput {
  ne: ID
  eq: ID
  le: ID
  lt: ID
  ge: ID
  gt: ID
  contains: ID
  notContains: ID
  between: [ID]
  beginsWith: ID
  attributeExists: Boolean
  attributeType: ModelAttributeTypes
  size: ModelSizeInput
}

input ModelSubscriptionStringInput {
  ne: String
  eq: String
  le: String
  lt: String
  ge: String
  gt: String
  contains: String
  notContains: String
  between: [String]
  beginsWith: String
  in: [String]
  notIn: [String]
}

input ModelSubscriptionIntInput {
  ne: Int
  eq: Int
  le: Int
  lt: Int
  ge: Int
  gt: Int
  between: [Int]
  in: [Int]
  notIn: [Int]
}

input ModelSubscriptionFloatInput {
  ne: Float
  eq: Float
  le: Float
  lt: Float
  ge: Float
  gt: Float
  between: [Float]
  in: [Float]
  notIn: [Float]
}

input ModelSubscriptionBooleanInput {
  ne: Boolean
  eq: Boolean
}

input ModelSubscriptionIDInput {
  ne: ID
  eq: ID
  le: ID
  lt: ID
  ge: ID
  gt: ID
  contains: ID
  notContains: ID
  between: [ID]
  beginsWith: ID
  in: [ID]
  notIn: [ID]
}

enum ModelAttributeTypes {
  binary
  binarySet
  bool
  list
  map
  number
  numberSet
  string
  stringSet
  _null
}

input ModelSizeInput {
  ne: Int
  eq: Int
  le: Int
  lt: Int
  ge: Int
  gt: Int
  between: [Int]
}

enum ModelSortDirection {
  ASC
  DESC
}

type ModelTodoConnection {
  items: [Todo]!
  nextToken: String
}

input ModelTodoFilterInput {
  id: ModelIDInput
  picture: ModelStringInput
  and: [ModelTodoFilterInput]
  or: [ModelTodoFilterInput]
  not: ModelTodoFilterInput
}

type Query {
  getTodo(id: ID!): Todo
  listTodos(filter: ModelTodoFilterInput, limit: Int, nextToken: String): ModelTodoConnection
}

input ModelTodoConditionInput {
  picture: ModelStringInput
  and: [ModelTodoConditionInput]
  or: [ModelTodoConditionInput]
  not: ModelTodoConditionInput
}

input CreateTodoInput {
  id: ID
  picture: AWSURL
}

input UpdateTodoInput {
  id: ID!
  picture: AWSURL
}

input DeleteTodoInput {
  id: ID!
}

type Mutation {
  createTodo(input: CreateTodoInput!, condition: ModelTodoConditionInput): Todo
  updateTodo(input: UpdateTodoInput!, condition: ModelTodoConditionInput): Todo
  deleteTodo(input: DeleteTodoInput!, condition: ModelTodoConditionInput): Todo
}

type Subscription {
  onCreateTodo: Todo @aws_subscribe(mutations: [\\"createTodo\\"])
  onUpdateTodo: Todo @aws_subscribe(mutations: [\\"updateTodo\\"])
  onDeleteTodo: Todo @aws_subscribe(mutations: [\\"deleteTodo\\"])
}

"
`;
